Beta3 crashes during analysis with leakcanary-android-process

### Description
When run in a separate process, Leakcanary crashes during heap analysis with:

```
 Process: com.airbnb.android.development:leakcanary, PID: 1125
 java.lang.ExceptionInInitializerError
 	at leakcanary.internal.ForegroundService.showForegroundNotification(ForegroundService.kt:50)
 	at leakcanary.internal.ForegroundService.onCreate(ForegroundService.kt:33)
 	at android.app.ActivityThread.handleCreateService(ActivityThread.java:3532)
 	at android.app.ActivityThread.access$1300(ActivityThread.java:199)
 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1666)
 	at android.os.Handler.dispatchMessage(Handler.java:106)
 	at android.os.Looper.loop(Looper.java:193)
 	at android.app.ActivityThread.main(ActivityThread.java:6669)
 	at java.lang.reflect.Method.invoke(Native Method)
 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
 Caused by: kotlin.UninitializedPropertyAccessException: lateinit property application has not been initialized
 	at leakcanary.internal.InternalLeakCanary.getApplication(InternalLeakCanary.kt:36)
 	at leakcanary.internal.Notifications.<clinit>(Notifications.kt:36)
 	... 11 more
```

I can get it to work if I manually add the app watcher during application creation with
```
if (LeakCanaryProcess.isInAnalyzerProcess(application)) {
      AppWatcher.manualInstall(application)
 }
```

### Steps to Reproduce
Follow the documentation to run heap analysis in a separate process by replacing the normal `leakcanary-android` dependency with `leakcanary-android-process`. 

1. Open Leaks activity
2. Import any hprof file for analysis
3. Note crash

The same crash also happens if the analysis happens because of routine usage of the app.

### Version Information

* LeakCanary version: v2-beta3
* Android OS version: 28
* Gradle version: 5.1

### Additional Information

N/A - Don't do any custom configuration, this is using the defaults and automatic setup that leak canary applies.

--- a/docs/fundamentals.md
+++ b/docs/fundamentals.md
@@ -235,3 +235,4 @@ LeakCanary is released as several libraries:
 * LeakCanary Android separate process
 	* Enables running LeakCanary separately from the main app process.
 	* Artifact id: `com.squareup.leakcanary:leakcanary-android-process` (as a replacement for `com.squareup.leakcanary:leakcanary-android`).
+	* See [Running the LeakCanary analysis in a separate process](recipes.md#running-the-leakcanary-analysis-in-a-separate-process).

--- a/docs/recipes.md
+++ b/docs/recipes.md
@@ -244,6 +244,21 @@ class DebugExampleApplication : ExampleApplication() {
 }
 ```
 
+## Running the LeakCanary analysis in a separate process
+
+LeakCanary runs in your main app process. LeakCanary 2 is optimized to keep memory usage low while analysing and runs in a background thread with priority `Process.THREAD_PRIORITY_BACKGROUND`. If you find that LeakCanary is still using too much memory or impacting the app process performance, you can configure it to run the analysis in a separate process.
+
+All you have to do is replace the `leakcanary-android` depedency with `leakcanary-android-process`:
+
+```groovy
+dependencies {
+  // debugImplementation 'com.squareup.leakcanary:leakcanary-android:${version}'
+  debugImplementation 'com.squareup.leakcanary:leakcanary-android-process:${version}'
+}
+```
+
+You can call [LeakCanaryProcess.isInAnalyzerProcess](/leakcanary/api/leakcanary-android-process/leakcanary/-leak-canary-process/is-in-analyzer-process/) to check if your Application class is being created in the LeakCanary process. This is useful when configuring libraries like Firebase that may crash when running in an unexpected process.
+
 ## Setting up LeakCanary for different product flavors
 
 You can setup LeakCanary to run in a specific product flavors of your app. For example, create:

--- a/leakcanary-android-core/src/main/java/leakcanary/LeakCanary.kt
+++ b/leakcanary-android-core/src/main/java/leakcanary/LeakCanary.kt
@@ -9,6 +9,7 @@ import shark.IgnoredReferenceMatcher
 import shark.LibraryLeakReferenceMatcher
 import shark.ObjectInspector
 import shark.ReferenceMatcher
+import shark.SharkLog
 
 /**
  * The entry point API for LeakCanary. LeakCanary builds on top of [AppWatcher]. AppWatcher
@@ -145,6 +146,10 @@ object LeakCanary {
    */
   @Volatile
   var config: Config = if (AppWatcher.isInstalled) Config() else InternalLeakCanary.noInstallConfig
+    set(value) {
+      field = value
+      SharkLog.d { "Updated LeakCanary.config to $value" }
+    }
 
   /**
    * Returns a new [Intent] that can be used to programmatically launch the leak display activity.

--- a/leakcanary-android-process/src/main/AndroidManifest.xml
+++ b/leakcanary-android-process/src/main/AndroidManifest.xml
@@ -21,13 +21,13 @@
     <service
         android:name="leakcanary.internal.HeapAnalyzerService"
         android:exported="false"
-        android:process=":leakcanary"
-        >
-      <intent-filter>
-        <action android:name="leakcanary.ANALYSIS_ACTION"/>
-      </intent-filter>
-    </service>
+        android:process=":leakcanary" />
 
+    <provider
+        android:name="leakcanary.internal.AppWatcherInstaller$LeakCanaryProcess"
+        android:authorities="${applicationId}.leakcanary-process.installer"
+        android:process=":leakcanary"
+        android:exported="false"/>
   </application>
 
 </manifest>

--- a/leakcanary-android/src/main/AndroidManifest.xml
+++ b/leakcanary-android/src/main/AndroidManifest.xml
@@ -20,13 +20,7 @@
   <application>
     <service
         android:name="leakcanary.internal.HeapAnalyzerService"
-        android:exported="false"
-        >
-      <intent-filter>
-        <action android:name="leakcanary.ANALYSIS_ACTION"/>
-      </intent-filter>
-    </service>
-
+        android:exported="false" />
   </application>
 
 </manifest>

--- a/leakcanary-object-watcher-android/src/main/AndroidManifest.xml
+++ b/leakcanary-object-watcher-android/src/main/AndroidManifest.xml
@@ -6,7 +6,7 @@
 
   <application>
     <provider
-        android:name="leakcanary.internal.AppWatcherInstaller"
+        android:name="leakcanary.internal.AppWatcherInstaller$MainProcess"
         android:authorities="${applicationId}.leakcanary-installer"
         android:exported="false"/>
   </application>

--- a/leakcanary-object-watcher-android/src/main/java/leakcanary/AppWatcher.kt
+++ b/leakcanary-object-watcher-android/src/main/java/leakcanary/AppWatcher.kt
@@ -1,9 +1,9 @@
 package leakcanary
 
 import android.app.Application
-import leakcanary.AppWatcher.config
 import leakcanary.AppWatcher.objectWatcher
 import leakcanary.internal.InternalAppWatcher
+import shark.SharkLog
 import java.util.concurrent.TimeUnit
 
 /**
@@ -61,6 +61,10 @@ object AppWatcher {
    */
   @Volatile
   var config: Config = if (isInstalled) Config() else Config(enabled = false)
+    set(value) {
+      field = value
+      SharkLog.d { "Updated AppWatcher.config to $value" }
+    }
 
   /**
    * The [ObjectWatcher] used by AppWatcher to detect retained objects.

--- a/leakcanary-object-watcher-android/src/main/java/leakcanary/internal/AppWatcherInstaller.kt
+++ b/leakcanary-object-watcher-android/src/main/java/leakcanary/internal/AppWatcherInstaller.kt
@@ -5,16 +5,32 @@ import android.content.ContentProvider
 import android.content.ContentValues
 import android.database.Cursor
 import android.net.Uri
-import shark.SharkLog
+import leakcanary.AppWatcher
 
 /**
  * Content providers are loaded before the application class is created. [AppWatcherInstaller] is
  * used to install [leakcanary.AppWatcher] on application start.
  */
-internal class AppWatcherInstaller : ContentProvider() {
+internal sealed class AppWatcherInstaller : ContentProvider() {
+
+  /**
+   * [MainProcess] automatically sets up the LeakCanary code that runs in the main app process.
+   */
+  internal class MainProcess : AppWatcherInstaller()
+
+  /**
+   * When using the `leakcanary-android-process` artifact instead of `leakcanary-android`,
+   * [LeakCanaryProcess] automatically sets up the LeakCanary code
+   */
+  internal class LeakCanaryProcess : AppWatcherInstaller() {
+    override fun onCreate(): Boolean {
+      super.onCreate()
+      AppWatcher.config = AppWatcher.config.copy(enabled = false)
+      return true
+    }
+  }
 
   override fun onCreate(): Boolean {
-    SharkLog.logger = DefaultCanaryLog()
     val application = context!!.applicationContext as Application
     InternalAppWatcher.install(application)
     return true

--- a/leakcanary-object-watcher-android/src/main/java/leakcanary/internal/InternalAppWatcher.kt
+++ b/leakcanary-object-watcher-android/src/main/java/leakcanary/internal/InternalAppWatcher.kt
@@ -57,6 +57,7 @@ internal object InternalAppWatcher {
   )
 
   fun install(application: Application) {
+    SharkLog.logger = DefaultCanaryLog()
     SharkLog.d { "Installing AppWatcher" }
     checkMainThread()
     if (this::application.isInitialized) {

